streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -49,10 +60,237 @@
+            this.config = config;
+            this.streamsMetrics = streamsMetrics;
+            this.stateDirectory = stateDirectory;
+            this.storeChangelogReader = storeChangelogReader;
+            this.time = time;
+            this.log = log;
+        }
+
+        public InternalTopologyBuilder builder() {
+            return builder;
+        }
+
+        public StateDirectory stateDirectory() {
+            return stateDirectory;
+        }
+
+        Collection<T> createTasks(final Consumer<byte[], byte[]> consumer,
+                                  final Map<TaskId, Set<TopicPartition>> tasksToBeCreated) {
+            final List<T> createdTasks = new ArrayList<>();
+            for (final Map.Entry<TaskId, Set<TopicPartition>> newTaskAndPartitions : tasksToBeCreated.entrySet()) {
+                final TaskId taskId = newTaskAndPartitions.getKey();
+                final Set<TopicPartition> partitions = newTaskAndPartitions.getValue();
+                final T task = createTask(consumer, taskId, partitions);
+                if (task != null) {
+                    log.trace("Created task {} with assigned partitions {}", taskId, partitions);
+                    createdTasks.add(task);
+                }
+
+            }
+            return createdTasks;
+        }
+
+        abstract T createTask(final Consumer<byte[], byte[]> consumer, final TaskId id, final Set<TopicPartition> partitions);
+
+        void close() {};
+    }
+
+    static class TaskCreator extends AbstractTaskCreator<StreamTask> {
+        private final ThreadCache cache;
+        private final Producer<byte[], byte[]> threadProducer;
+        private final Map<TaskId, Producer<byte[], byte[]>> taskProducers;
+        private final KafkaClientSupplier clientSupplier;
+        private final Sensor createTaskSensor;
+
+        TaskCreator(final InternalTopologyBuilder builder,
+                    final StreamsConfig config,
+                    final StreamsMetricsImpl streamsMetrics,
+                    final StateDirectory stateDirectory,
+                    final ChangelogReader storeChangelogReader,
+                    final ThreadCache cache,
+                    final Time time,
+                    final KafkaClientSupplier clientSupplier,
+                    final Producer<byte[], byte[]> threadProducer,
+                    final Map<TaskId, Producer<byte[], byte[]>> taskProducers,
+                    final Logger log) {
+            super(
+                builder,
+                config,
+                streamsMetrics,
+                stateDirectory,
+                storeChangelogReader,
+                time,
+                log);
+
+            this.threadProducer = threadProducer;
+            this.taskProducers = taskProducers;
+
+            this.cache = cache;
+            this.clientSupplier = clientSupplier;
+
+            this.createTaskSensor = ThreadMetrics.createTaskSensor(Thread.currentThread().getName(), streamsMetrics);
```

Comment:
Minor change: We don't pass in `threadId` any longer but just call `Thread.currentThread.getName()` instead -- let me know what you think about it -- seems simpler to me to avoid passing around unnecessary parameters